Correct handling of timestamps belonging to leap days

**Describe the bug**
When ksql encounters a timestamp belonging to a leap day, it currently fails with the following

```
Caused by: java.time.DateTimeException: Invalid date 'DayOfYear 366' as '1970' is not a leap year
at java.time.LocalDate.ofYearDay(LocalDate.java:290)
at java.time.LocalDate.withDayOfYear(LocalDate.java:1118)
at java.time.LocalDate.with(LocalDate.java:1029)
at java.time.LocalDateTime.with(LocalDateTime.java:970)
at java.time.ZonedDateTime.with(ZonedDateTime.java:1311)
at io.confluent.ksql.util.timestamp.StringToTimestampParser.parseZoned(StringToTimestampParser.java:75)
at io.confluent.ksql.util.timestamp.StringToTimestampParser.parse(StringToTimestampParser.java:58)
at io.confluent.ksql.util.timestamp.StringToTimestampParser.parse(StringToTimestampParser.java:48)
at io.confluent.ksql.util.timestamp.StringTimestampExtractor.extract(StringTimestampExtractor.java:45)
at org.apache.kafka.streams.processor.internals.RecordQueue.updateHead(RecordQueue.java:177)
```


**To Reproduce**

Able to reproduce this with a small test case

```
@Test
  public void testEscalation2522() {
    // Given
    final String format = "yyyy-MM-dd'T'HH:mm:ss.SSSX";
    final String timestamp = "2012-12-31T23:59:58.660Z";

    // When
    final long ts = new StringToTimestampParser(format).parse(timestamp);

    // Then

  }
```

**Expected behavior**

Valid epoch timestamp is returned. 

**Actual behaviour**

```
java.time.DateTimeException: Invalid date 'DayOfYear 366' as '1970' is not a leap year

	at java.time.LocalDate.ofYearDay(LocalDate.java:290)
	at java.time.LocalDate.withDayOfYear(LocalDate.java:1118)
	at java.time.LocalDate.with(LocalDate.java:1029)
	at java.time.LocalDateTime.with(LocalDateTime.java:970)
	at java.time.ZonedDateTime.with(ZonedDateTime.java:1311)
	at io.confluent.ksql.util.timestamp.StringToTimestampParser.parseZoned(StringToTimestampParser.java:75)
	at io.confluent.ksql.util.timestamp.StringToTimestampParser.parse(StringToTimestampParser.java:58)
	at io.confluent.ksql.util.timestamp.StringToTimestampParser.parse(StringToTimestampParser.java:48)
	at io.confluent.ksql.util.timestamp.StringToTimestampParserTest.testEscalation2522(StringToTimestampParserTest.java:66)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:33)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:230)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:58)
```

**Additional context**
Add any other context about the problem here.

--- a/ksqldb-common/src/main/java/io/confluent/ksql/util/timestamp/StringToTimestampParser.java
+++ b/ksqldb-common/src/main/java/io/confluent/ksql/util/timestamp/StringToTimestampParser.java
@@ -32,6 +32,7 @@ public class StringToTimestampParser {
 
   private static final Function<ZoneId, ZonedDateTime> DEFAULT_ZONED_DATE_TIME =
       zid -> ZonedDateTime.of(1970, 1, 1, 0, 0, 0, 0, zid);
+  private static final long LEAP_DAY_OF_THE_YEAR = 366;
 
   private final DateTimeFormatter formatter;
 
@@ -72,7 +73,16 @@ public class StringToTimestampParser {
           throw new KsqlException(
               "Unsupported temporal field in timestamp: " + text + " (" + override + ")");
         }
-        resolved = resolved.with(override, parsed.getLong(override));
+
+        final long value = parsed.getLong(override);
+        if (override == ChronoField.DAY_OF_YEAR && value == LEAP_DAY_OF_THE_YEAR) {
+          if (!parsed.isSupported(ChronoField.YEAR)) {
+            throw new KsqlException("Leap day cannot be parsed without supplying the year field");
+          }
+          // eagerly override year, to avoid mismatch with epoch year, which is not a leap year
+          resolved = resolved.withYear(parsed.get(ChronoField.YEAR));
+        }
+        resolved = resolved.with(override, value);
       }
     }
 

--- a/ksqldb-common/src/test/java/io/confluent/ksql/util/timestamp/StringToTimestampParserTest.java
+++ b/ksqldb-common/src/test/java/io/confluent/ksql/util/timestamp/StringToTimestampParserTest.java
@@ -3,6 +3,7 @@ package io.confluent.ksql.util.timestamp;
 import static org.hamcrest.MatcherAssert.assertThat;
 import static org.hamcrest.Matchers.is;
 
+import io.confluent.ksql.util.KsqlException;
 import java.time.ZoneId;
 import java.time.ZonedDateTime;
 import org.hamcrest.Description;
@@ -21,6 +22,9 @@ public class StringToTimestampParserTest {
 
   private static final ZonedDateTime FIFTH_OF_NOVEMBER =
       ZonedDateTime.of(1605, 11, 5, 0, 0, 0, 0, ZID);
+
+  private static final ZonedDateTime NEW_YEARS_EVE_2012 =
+      ZonedDateTime.of(2012, 12, 31, 23, 59, 58, 660000000, ZID);
 
   @Test
   public void shouldParseBasicLocalDate() {
@@ -145,6 +149,29 @@ public class StringToTimestampParserTest {
 
     // Then
     assertThat(ts, is(sameInstant(FIFTH_OF_NOVEMBER.withHour(10))));
+  }
+
+  @Test
+  public void shouldParseLeapDay() {
+    // Given
+    final String format = "yyyy-MM-dd'T'HH:mm:ss.SSS";
+    final String timestamp = "2012-12-31T23:59:58.660";
+
+    // When
+    final ZonedDateTime ts = new StringToTimestampParser(format).parseZoned(timestamp, ZID);
+
+    // Then
+    assertThat(ts, is(sameInstant(NEW_YEARS_EVE_2012)));
+  }
+
+  @Test(expected = KsqlException.class)
+  public void shouldThrowErrorForLeapDayWithoutYear() {
+    // Given
+    final String format = "DDD";
+    final String timestamp = "366";
+
+    // When
+    new StringToTimestampParser(format).parseZoned(timestamp, ZID);
   }
 
   @Test

